### PR TITLE
improve the structure of the document

### PR DESCRIPTION
**Pull request by you-n-g**

### CODE REVIEW

File: binding/python/README.md [hunk 1]
```diff
@@ -1,225 +1,21 @@
 
 
-# Requirements
+# Installation
 
 ## On linux
-Please followed the [README](../../README.md#build) and have build and install multiverso successfully.
+Please followed the [README](../../README.md#build) to build and install multiverso.
```
Comment (huxuan) [inline binding/python/README.md]: "Please follow"

Comment (you-n-g) [inline binding/python/README.md]: Amended

File: binding/python/README.md [hunk 2]
```diff
@@ -1,225 +1,21 @@
...
 
 ## On windows
-You need MSBuild.exe installed and your system can find it in the $PATH. Then you should run [build_dll.bat](../../src/build_dll.bat) to build the .dll file and install the .dll. Multiverso doesn't have auto-installer for windows now, you have to copy the .dll to either system $PATH or the multiverso package folder.
+You need MSBuild.exe installed and make sure your system can find it in the $PATH. Then you should run [build_dll.bat](../../src/build_dll.bat) to build the .dll file and install the .dll. Multiverso doesn't have auto-installer for windows now, you have to copy the .dll to either system $PATH or the multiverso package folder.
```
Comment (huxuan) [inline binding/python/README.md]: Add quote for `$PATH` (like this). and I suggest to use lists instead of paragraph description, something like.
```
1. Have `MSBuild.exe` installed and make it available in the `$PATH`.
2. Run [build_dll.bat](../../src/build_dll.bat) to build the `.dll` Dynamic Link Library file.
3. Copy/Install the Dynamic Link Library file to the system `$PATH` or multiverso package folder.
```

File: binding/python/README.md [hunk 3]
```diff
@@ -1,225 +1,21 @@
...
-|1M-4G-1S | 1 | 4 | 4 | 1 | 92.15 % | 29.38|
-|1M-4G-3S | 1 | 4 | 4 | 3 | 89.61 % | 27.46|
-
-![accuracy_epoch](./docs/imgs/accuracy_epoch.png)
-![accuracy_time](./docs/imgs/accuracy_time.png)
-![time_epoch](./docs/imgs/time_epoch.png)
+# Documentation
+* [Experiments](./docs/EXPERIMENTS.md)
+* [Tutorial](./docs/TUTORIAL.md)
+* All the api documents are written as docstrings in the python source.
```
Comment (huxuan) [inline binding/python/README.md]: "API documents are written as docstrings in the python source code."

File: binding/python/docs/EXPERIMENTS.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# Experiments
+
+Here is the result of running [Deep_Residual_Learning_CIFAR-10](../examples/theano/lasagne/Deep_Residual_Learning_CIFAR-10.py)
```
Comment (huxuan) [inline binding/python/docs/EXPERIMENTS.md]:
```
## Codebase

[Deep_Residual_Learning_CIFAR-10](../examples/theano/lasagne/Deep_Residual_Learning_CIFAR-10.py)
```

Maybe in a different place like below `Task Description`.

File: binding/python/docs/EXPERIMENTS.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+|||
+| -------- |:--------:|
+|Hosts|1|
+|GPU|GeForce GTX TITAN X * 4|
+|CPU|Intel(R) Core(TM) i7-5960X CPU @ 3.00GHz  * 1|
+|Memory| 128GB |
+
+
+## Theano settings
+Here is the content of `~/.theanorc`
```
Comment (huxuan) [inline binding/python/docs/EXPERIMENTS.md]: "Configuration in `~/.theanorc`" or maybe combine with the previous line as something like "## Theano Settings (`~/.theanorc`)"

File: binding/python/docs/EXPERIMENTS.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+cnmem = 1
+```
+
+## About the Model
+|||
+| :---- | -----: |
+|Total epoch|82|
+|Batch size|128|
+|Depth|32|
+|Learning rate change schedule|Initialized as 0.1, Changed to 0.01 from epoch 41, Changed to 0.001 from epoch 61|
```
Comment (huxuan) [inline binding/python/docs/EXPERIMENTS.md]: "Change to" for both existences.

File: binding/python/docs/EXPERIMENTS.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+|number of parameters in model|    464,154|
```
Comment (huxuan) [inline binding/python/docs/EXPERIMENTS.md]: "Number of parameters"

File: binding/python/docs/EXPERIMENTS.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+
+Clarification
+- An epoch represents all the processes divide all the data equally and go through them once together.
```
Comment (huxuan) [inline binding/python/docs/EXPERIMENTS.md]: "An epoch means divide all the training data evenly to all workers and each worker goes through the assigned data once."

File: binding/python/docs/EXPERIMENTS.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+- A barrier is used at the end of every epoch.
```
Comment (huxuan) [inline binding/python/docs/EXPERIMENTS.md]: "each epoch"

File: binding/python/docs/EXPERIMENTS.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+- This experiment didn't use warm start in ASGD.
```
Comment (huxuan) [inline binding/python/docs/EXPERIMENTS.md]: "This experiments doesn't use warm start in ASGD."

File: binding/python/docs/EXPERIMENTS.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+- The time to load the data is not considered in the time of the experiment.
+
+
+# The results
+4 experiments have been run. The configuration and the results of each configuration are listed below.
```
Comment (huxuan) [inline binding/python/docs/EXPERIMENTS.md]: The results of 4 experiments with different configurations are shown as following.

File: binding/python/docs/EXPERIMENTS.md [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+|Short Name | With multiverso | number of Process | number of GPU | Sync every X minibatches |  Best model validation accuracy | Time per epoch / s |
```
Comment (huxuan) [inline binding/python/docs/EXPERIMENTS.md]: "Number of process(es)"
"Number of GPU(s)"

File: binding/python/docs/TUTORIAL.md [hunk 13]
```diff
@@ -0,0 +1,144 @@
+# How to write python code with multiverso
+1. You could start with the [test example](../multiverso/tests/test_multiverso.py) to learn the basic use of multiverso apis.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: "Start with the" "basic usage"

File: binding/python/docs/TUTORIAL.md [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+2. After understanding the basic use of multiverso apis, you can test and run the [examples](../examples/).  The original code links are written at the beginning of them. You can compare the multiverso version with the original ones to find the differences. The places need to be modified to use multiverso are all inserted comments like `# MULTIVERSO: XXX`
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]:
```
Try the [examples](../examples/): The link of the original version are listed at the header part. All the modifications are commented with `# MULTIVERSO: XXX` and you can compare them with the original one.
```

File: binding/python/docs/TUTORIAL.md [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+
+Here is a typical example of python code with multvierso.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: "Here is a typical usage of multiverso python binding."

File: binding/python/docs/TUTORIAL.md [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+# Init multiverso.
+mv.init()
+# Get total number of workers.
+print mv.workers_num()
+# Get the id for current worker.
+print mv.worker_id()
+# if the worker is master worker.
+print mv.is_master_worker()
+
+# Here is your code to create tables, train models and sync values.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: "Here is the code for training." (It's a little strange to introduce tables and sync concepts here.)

File: binding/python/docs/TUTORIAL.md [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+
+# Shutdown multiverso
+mv.shutdown()
+```
+
+## About the master worker
+Some things only need one worker process, such as validation, outputting the result, initializing the parameters and so on. So we mark the worker 0 as the master worker to finish these things. You can use `mv.is_master_worker` to distinguish the master worker from others.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: Something should only be done in one specific worker, such as validation, outputting the results, initializing the parameters and so on. So you can benefit from `mv.is_master_worker` api to mark worker with worker_id 0 as the master one.

File: binding/python/docs/TUTORIAL.md [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+    # Set a barrier for other workers to wait.
+    mv.barrier()
+else:
+    # Wait the master worker to finish setting.
+    mv.barrier()
+    # Get the initial model from the server.
+    params = tbh.get()
+```
+
+There are similar strategies in `theano_ext.sharedvar` and `lasagne_ext.param_manager` when initializing values. They are already implemented in the constructors.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: Similar strategies also apply to `theano_ext.sharedvar` and `lasagne_ext.param_manager` during initialization and it's already implemented in the constructors.

File: binding/python/docs/TUTORIAL.md [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+
+
+
+# How to use multiverso in theano
+First, make sure you have understood `mv.init()`, `mv.shutdown()` and `mv.barrier()` mentioned in last section.  You should add the same functions in your theano python script.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: First, Similarly, add `mv.init()`, `mv.shutdown()` and `mv.barrier()` mentioned above in your codebase.

File: binding/python/docs/TUTORIAL.md [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+        (n_in, n_out),
+        dtype=theano.config.floatX
+    ),
+    name='W',
+    borrow=True
+)
+
+# build the model
+
+# train the modle
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: model

File: binding/python/docs/TUTORIAL.md [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+
+# When you are ready to add the delta of the variable to parameter server and sync the latest value, you can run this function
+W.mv_sync()
+
+
+# If you want to sync all variables created by `sharedvar.mv_shared`, you can use this function
+sharedvar.sync_all_mv_shared_vars()
+```
+
+`mv_shared` is just a wrapper of `theano.shared`. It acts same as `theano.shared`, but it give you much more convenient interface to sync values.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: It acts the same as `theano.shared` while making it more convenient to sync values.

File: binding/python/docs/TUTORIAL.md [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+
+If your program will run in multi-process and you want to initialize your parameters, you should initialize your shared variables with the value you want only in one process and initialize them with zero in other processes.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: For initializing parameters in multi-processes/hosts programs, initialization is only necessary in the master worker. (Is the master worker right?)

File: binding/python/docs/TUTORIAL.md [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+
+
+If you don't use shared variables to store and update the parameters, you can still use the `add` and `get` functions to sync parameters.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: `add` and `get` can also be used to sync parameters even you don't use shared variables.

File: binding/python/docs/TUTORIAL.md [hunk 24]
```diff
@@ -0,0 +1,144 @@
...
+
+
+# How to use multiverso in lasagne
+First, make sure you have understood the last section.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: I would rather have this line removed.

File: binding/python/docs/TUTORIAL.md [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+
+Lasagne provides many functions to help you build models more easily. Multiverso python binding provides a manager to make managing and synchronizing the parameters in Lasagne more easily.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: Lasagne provides many functions to build models while multiverso python binding also provides a manager to make synchronization in Lasagene more easily.

File: binding/python/docs/TUTORIAL.md [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+
+You can write code like this to manage your parameters.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: A typical usage of managing the parameters as shown as below.

File: binding/python/docs/TUTORIAL.md [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+```
+from multiverso.theano_ext.lasagne_ext import param_manager
+
+network = build_model()  # build_model is a function you implement to build model
+
+# The MVNetParamManager will initialize the parameters and sync them with
+# parameter server
+mvnpm = param_manager.MVNetParamManager(network)
+
+# training the model
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: Train the model.

File: binding/python/docs/TUTORIAL.md [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+
+# When you are ready to add the delta of the variable in this model to the parameter server and get the latest value, you can run this function
+mvnpm.sync_all_param()
+```
+
+
+# Run your multiverso program with multi-process
+Here is an example of running logistic regression with multi-process.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: Command to run logistic regression with 4 processes.

File: binding/python/docs/TUTORIAL.md [hunk 29]
```diff
@@ -0,0 +1,144 @@
...
+```
+mpirun -np 4 python ./examples/theano/logistic_regression.py
+```
+
+
+# How to use multi-GPU in theano with multiverso
+You need multiple GPUs in your server and have installed the [CUDA backend](http://deeplearning.net/software/theano/tutorial/using_gpu.html#cuda).
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: You need multiple GPUs on the server and have the [CUDA backend](....) installed.

File: binding/python/docs/TUTORIAL.md [hunk 30]
```diff
@@ -0,0 +1,144 @@
...
+
+First, make sure you have read [this section](http://deeplearning.net/software/theano/install.html#using-the-gpu) and understand how to configure which GPU will be used.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: First, look through [this section]() to understand the configuration of which GPU will be used.

File: binding/python/docs/TUTORIAL.md [hunk 31]
```diff
@@ -0,0 +1,144 @@
...
+
+With multiverso, your program will run in multiple processes.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: Second, run the program with multiverso in multiple processes. (Or maybe make them as a numbered list?

File: binding/python/docs/TUTORIAL.md [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+
+Here is an example to make different GPUs used in different processes.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: "make different process use different GPUs" or "have different GPUs used in different processes".

File: binding/python/docs/TUTORIAL.md [hunk 33]
```diff
@@ -0,0 +1,144 @@
...
+In this example, the i-th worker will use the i-th gpu. You need to add code like this before `import theano`.
```
Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: "GPU" (just to make it consistent)

**Comment by huxuan** [comment]
Done for comments. 

Some general comments:

- All comments / docstrings should be a complete sentence, like uppercase for the first character, period punctuation or other ones at the end. (This should also apply to commit message and all other similar places.
- I think we should keep the line width limitation even for docs. Especially for the cases of two completed sentences in one line will make it hard to review and comment.

File: binding/python/docs/TUTORIAL.md (hunk 16 above)
Comment (you-n-g) [inline binding/python/docs/TUTORIAL.md]: I think the README of binding is responsible for telling users how to use binding and extra concept introduced by binding.   The  sync and table concept should be introduced somewhere else.

Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: From my experience I prefer to look through tutorial first rather than api documents directly. So we should suppose users are newbies when first come to the tutorial. What do you think of it?

File: binding/python/docs/TUTORIAL.md (hunk 22 above)
Comment (you-n-g) [inline binding/python/docs/TUTORIAL.md]: I think we could delete this line now. Because we will implement this in our constructor and won't expose it to users.w

Comment (huxuan) [inline binding/python/docs/TUTORIAL.md]: Yep.

File: binding/python/docs/TUTORIAL.md (hunk 25 above)
Comment (you-n-g) [inline binding/python/docs/TUTORIAL.md]: I think seperate the two sentences will be better. 
Perhaps we can change the first one  to `Lasagne provides many functions to build models in theano` to make it concise

**Comment by you-n-g** [comment]
The documents are updated.
Thanks  again for such detailed comments !!!!!!
